pkg/util/provider/machinecontroller/machine.go

```diff
@@ -234,6 +234,33 @@ func (c *controller) addNodeToMachine(obj interface{}) {
 }
 
 func (c *controller) updateNodeToMachine(oldObj, newObj interface{}) {
+
+	node := newObj.(*corev1.Node)
+	if node == nil {
+		klog.Errorf("Couldn't convert to node from object")
+		return
+	}
+
+	// check for the TriggerDeletionByMCM annotation on the node object
+	// if it is present then mark the machine object for deletion
+	if value, ok := node.Annotations[machineutils.TriggerDeletionByMCM]; ok && value == "true" {
+
+		klog.Infof("Node %s is annotated to trigger deletion by MCM.", node.Name)
+		machine, err := c.getMachineFromNode(node.Name)
+		if err != nil {
+			klog.Errorf("Couldn't fetch machine %s, Error: %s", machine.Name, err)
+			return
+		}
+
+		if machine.DeletionTimestamp == nil {
+			if err := c.controlMachineClient.Machines(c.namespace).Delete(context.Background(), machine.Name, metav1.DeleteOptions{}); err != nil {
+				klog.Errorf("Machine object %s backing the node %s could not be marked for deletion.", machine.Name, node.Name)
+				return
+			}
+			klog.Infof("Machine object %s backing the node %s marked for deletion.", machine.Name, node.Name)
+		}
+	}
+
 	c.addNodeToMachine(newObj)
```

Comment:
I have tested the code locally , it runs as expected
Just one minor pick, I see the log `Node %s is annotated to trigger deletion by MCM.` repeated multiple times in the mc-logs.
The possible reason for it might be `updateNodeToMachine` is getting called for any other update on the node obj by other controllers during deletion.
We can mitigate this by moving the log to `if machine.DeletionTimestamp == nil` code block.

Reply:
Good catch! Updated in the next commit